### DISCUSSION TITLE
Vercel Python runtime (`"@vercel/python"`) build image does not have OpenSSL installed

### DISCUSSION THREAD

**Discussion by zsiegel92** [vercel/community#5554] [Bug Report]
### Page to Investigate

https://vercel.com/docs/deployments/build-image

### Steps to Reproduce

1. Use a Python runtime in a Vercel app:

```json
{
  "$schema": "https://openapi.vercel.sh/vercel.json",
  "builds": [
    {
      "src": "api/py/index.py",
      "use": "@vercel/python"
    }
  ],
  "routes": [{ "src": "/py/(.*)", "dest": "api/py/index.py" }]
}
```

2. Use a Python package that relies on system `openssl` command. In my case, Python's Prisma packge calls `openssl version -v` using `subprocess` [here](https://github.com/RobertCraigie/prisma-client-py/blob/3f16ebabbb41651aaf075f629ee5fb5d5cffb7aa/src/prisma/binaries/platform.py#L60-L69) as soon as a Prisma connection is opened (`prisma = Prisma();
    await prisma.connect()`).

3. Get an error in Vercel runtime logs: "No such file or directory: 'openssl'".

## Should OpenSSL be installed? The docs say yes.

[Vercel's documentation on build images](https://vercel.com/docs/deployments/build-image) states:

> Vercel uses [Amazon Linux 2](https://aws.amazon.com/amazon-linux-2/) as a base image for all builds and several pre-installed packages. The following packages are pre-installed in the build image with yum, the default package manager for Amazon Linux 2.
...

The following are listed among these "pre-installed" dependencies:

> `openssl`, `openssl-devel`, `openssl-libs`

## Am I confident OpenSSL is indeed not installed? Yes.

First of all, `openssl` is not available on the PATH, which is why `openssl version -v` fails (in Prisma's Python package, called via `subprocess`). Some debugging shows that there is no executable `openssl` in `/usr/bin`.

Furthermore, per [the Vercel build image documentation's suggestion](https://vercel.com/docs/deployments/build-image), I ran the base image locally using the following:

```sh
docker run --rm -it amazonlinux:2.0.20191217.0 sh
```
and used `yum install openssl` to confirm that it did in fact result in an executable at `/usr/bin/openssl`.

So, when my Python function executes, `openssl` is not on the PATH, and `/usr/bin/openssl` is missing. I am confident `openssl` is not installed, contradicting the Vercel documentation.

Please update the `"@vercel/python"` build image!

Thank you.

**Comment by swarnava** [comment]
Hi there,

You should no longer need to define `@vercel/python` in your vercel.json. Here is a sample project: https://github.com/vercel-support/vercel-fastapi/blob/main/vercel.json

**Reply by zsiegel92** [reply]
@swarnava thanks for this response.

I changed my `vercel.json` to use the `"functions"` parameter (and to omit an explicit "runtime"), and **I still experience the same issue**.

The `openssl` command is not available on the container where the endpoint runs! (Despite, as I mentioned above, Vercel explicitly stating that it should be available.)

Note that I was able to get my project to work with `vercel dev`, and when running in that way `openssl` is available.

@swarnava can you please confirm with someone on the team that maintains the Python runtime that OpenSSL is installed?

For what it's worth, this is my `vercel.json` file:

```json
{
  "$schema": "https://openapi.vercel.sh/vercel.json",
  "functions": {
    "api/py/index.py": {
      "maxDuration": 10
    }
  },
  "routes": [
    {
      "src": "/api/py(?:/(.*))?",
      "dest": "api/py/index.py"
    }
  ]
}
```

**Comment by zsiegel92** [comment]
@swarnava this conversation went off-track regarding my `vercel.json` file. I've gotten Python 3.9 applications working on Vercel, and this issue really only concerns the absence of OpenSSL installed in the Python 3.9 container. Vercel's documentation claims that OpenSSL is part of the Python 3.9 images on Vercel, but I documented above that the containers do not in fact seem to have OpenSSL.

Please do let me know whether/how I can document any debugging that could help Vercel confirm and resolve this bug.

Thank you!
